File: src/game/utils/asset.lua
local Asset = {}

local _imageCache = {}
local _soundCache = {}

local function get(cache, loader, file)
  if (cache[file]) then
    return cache[file]
  else
    local asset = loader(file)
    cache[file] = asset

    return asset
  end
end

function Asset.getImage(file)
  local image = get(_imageCache, love.graphics.newImage, file)
  image:setWrap('repeat', 'repeat')

  return image
end

function Asset.getSound(file)
  local sound = get(_soundCache, love.audio.newSource, file)

  return sound
end

return Asset

File: client2/src/game/utils/asset.ts
const imageCache: Map<string, Image> = new Map()
const audioCache: Map<string, Source> = new Map()

export const loadImage = (filepath: string) => {
  if (imageCache.get(filepath)) {
    return imageCache.get(filepath)
  } else {
    const image = love.graphics.newImage(filepath)
    image.setWrap('repeat', 'repeat')
    imageCache.set(filepath, image)

    return image
  }
}

export const loadAudio = (filepath: string) => {
  if (audioCache.get(filepath)) {
    return audioCache.get(filepath)
  } else {
    const audio = love.audio.newSource(filepath, 'static')
    audioCache.set(filepath, audio)

    return audio
  }
}
